### PR TITLE
SNOW-610866 Add support for SQLAlchemy’s dialect tests

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-610866 add Add support for SQLAlchemy’s dialect tests 

**This PR does the following tasks:**
1. import the sqlalchemy test suite into a standalone directory
2. add pytest.init to ignore test suites by default because it conflicts existing snowflake specific tests
3. instructions on how to run sqlalchemy test suites in standalone directory
4. update tox to be 2-steps tests process
5. skip tests in test suites failing due to
  - snowflake db does not support
  - snowflake sqlalchemy does not implement
  - needs more investigation

the test_suites.py contains all the failing tests skipped with detailed information

- next steps:
  - create issue in the sqlalchemy repo about the pytest plugin encountered
  - investigate the failing tests, either skip + reason or it's something needed to be fixed in snowflake-sqlalchemy
  - check through the requirements.py to see if there's any other requirement we should turn on/off
  - create separate PRs for the fix

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/sqlalchemy/snowdialect.py [hunk 1]
```diff
@@ -165,6 +165,9 @@ class SnowflakeDialect(default.DefaultDialect):
     # and denormalize_name() must be provided.
     requires_name_normalize = True
 
+    # READ_COMMITTED is the only supported isolation level
+    default_isolation_level = "READ_COMMITTED"
```
Comment (sfc-gh-aling) [inline src/snowflake/sqlalchemy/snowdialect.py:169]: this is one sample fix (or should not be a fix) to set the default isolation level of our dialect

File: src/snowflake/sqlalchemy/snowdialect.py [hunk 2]
```diff
@@ -304,8 +307,10 @@ def _get_schema_primary_keys(self, connection, schema, **kw):
             )
         )
         ans = {}
+        key_sequence_order_map = defaultdict(list)
```
Comment (sfc-gh-aling) [inline src/snowflake/sqlalchemy/snowdialect.py:310]: this is another sample fix which the order of key_sequence should be kept

File: tests/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env python
```
Comment (sfc-gh-aling) [inline tests/__init__.py:1]: @sfc-gh-mkeller any concern on turning the tests into a module?

File: tests/sqlalchemy_test_suite/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# SQLAlchemy Compliance Tests
+
+SQLAlchemy offers tests to test SQLAlchemy dialects work properly. This directory applies these tests
+to the Snowflake SQLAlchemy dialect.
+
+**Please be aware that the test suites are not collected in pytest by default** -- the directory is ignored in `pytest.ini`.
+This is because importing sqlalchemy pytest plugin will result in Snowflake SQLAlchemy dialect specific tests not
+being collected, also the Python parser can not load options correctly from setup.cfg or from pytest plugin steps.
+However, running from a separate
```
Comment (sfc-gh-mkeller) [inline tests/sqlalchemy_test_suite/README.md:9]: ?

File: tox.ini [hunk 5]
```diff
@@ -38,6 +38,10 @@ commands = pytest \
            --cov "snowflake.sqlalchemy" \
            --junitxml {toxworkdir}/junit.{envname}.xml \
            {posargs:tests}
+           pytest {env:SNOWFLAKE_PYTEST_OPTS:} \
+           --cov "snowflake.sqlalchemy" --cov-append \
+           --junitxml {toxworkdir}/junit.{envname}.xml \
+           tests/sqlalchemy_test_suite
```
Comment (sfc-gh-mkeller) [inline tox.ini:44]:
```suggestion
           {posargs:tests/sqlalchemy_test_suite}
```

File: tests/sqlalchemy_test_suite/__init__.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env python
+#
+# Copyright (c) 2012-2022 Snowflake Computing Inc. All right reserved.
```
Comment (sfc-gh-mkeller) [inline tests/sqlalchemy_test_suite/__init__.py:3]: You should update all of the copyrights. You could copy the pre-commit hook that we use in the Python connector to make migrating these simple in the future

File: tests/sqlalchemy_test_suite/test_suite.py [hunk 7]
```diff
@@ -0,0 +1,499 @@
+#!/usr/bin/env python
```
Comment (sfc-gh-mkeller) [inline tests/sqlalchemy_test_suite/test_suite.py:1]: These should not be necessary, please remove them everywhere

File: tests/sqlalchemy_test_suite/test_suite.py [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    def test_select_direct(self, connection):
+        """
+        >       assert a == b, msg or "%r != %r" % (a, b)
+        E       AssertionError: '2012-10-15 12:57:18' != datetime.datetime(2012, 10, 15, 12, 57, 18)
+        """
+        pass
+
+
+class DateTimeMicrosecondsTest(_DateTimeMicrosecondsTest):
+    @pytest.mark.skip("need investigation")
```
Comment (sfc-gh-mkeller) [inline tests/sqlalchemy_test_suite/test_suite.py:258]: Are you not going to investigate these in this PR? If not then please file a follow-up task!

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: pytest.ini [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# pytest.ini
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply pytest.ini]: file removed

File: pytest.ini [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+[pytest]
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply pytest.ini]: file removed

File: setup.cfg [hunk 11]
```diff
@@ -74,3 +74,11 @@ development =
     numpy
 pandas =
     snowflake-connector-python[pandas]<3.0.0
+
+[tool:pytest]
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply setup.cfg]: moved to tox.ini

File: tests/sqlalchemy_test_suite/README.md (hunk 4 above)
Comment (sfc-gh-aling) [reply tests/sqlalchemy_test_suite/README.md]: content updated

File: tests/sqlalchemy_test_suite/test_suite.py (hunk 7 above)
Comment (sfc-gh-aling) [reply tests/sqlalchemy_test_suite/test_suite.py]: I've put the changes in another PR: https://github.com/snowflakedb/snowflake-sqlalchemy/pull/306

File: tests/sqlalchemy_test_suite/__init__.py (hunk 6 above)
Comment (sfc-gh-aling) [reply tests/sqlalchemy_test_suite/__init__.py:2]: I've put the changes in another PR: https://github.com/snowflakedb/snowflake-sqlalchemy/pull/306

File: src/snowflake/sqlalchemy/requirements.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python
+#
+# Copyright (c) 2012-2022 Snowflake Computing Inc. All right reserved.
+#
+
+from sqlalchemy.testing import exclusions
+from sqlalchemy.testing.requirements import SuiteRequirements
+
+# TODO: check through the requirement list to determine which should be turned on/off
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/sqlalchemy/requirements.py:8]: follow-up task created: https://snowflakecomputing.atlassian.net/browse/SNOW-619615

**Review by sfc-gh-mkeller — APPROVED** [review]

**Pull request merged**